### DISCUSSION TITLE
Possible issues with new devtools-archlinuxcn

### DISCUSSION THREAD

**Discussion by yan12125** [archlinuxcn/repo#2815] [General]
devtools-archlinuxcn on the build server is upgraded to the latest version. Upstream devtools introduces several changes that may fail builds or cause runtime crashes.

### LTO is now enabled by default

This change is from [1]. Some known issues:

* May break object files or static libs in packages: https://archlinux.org/todo/lto-fat-objects/
* Go with CGO may be broken https://github.com/golang/go/issues/43505
* There are reports about broken LTO with Rust + C libraries. I don't understand that and didn't find a upstream report, though.

`options+=(!lto)` is generally a viable workaround for such issues.

### New security flags for compilers

This change is from [2]. Some relevant issues:

* `-Werror=format-security`: may cause compile errors like

> error: format not a string literal and no format arguments [-Werror=format-security]

(From https://buildarm.archlinuxcn.org/imlonghao-api/pkg/gdb-git/log/1654708826)
 
The fix is refactoring calls to printf-like functions.

* `-Wp,-D_GLIBCXX_ASSERTIONS`: programs may abort with `Assertion 'xxx' failed`. The proper fix, identifying incorrect usage in the code and then fixing it (ex: [3]), is generally difficult. As a result, many packagers simply remove the flag (ex: chromium, electron and ispc [4])

[1] https://gitlab.archlinux.org/archlinux/rfcs/-/blob/master/rfcs/0004-lto-by-default.rst
[2] https://gitlab.archlinux.org/archlinux/rfcs/-/blob/master/rfcs/0003-buildflags.rst
[3] https://github.com/pytorch/pytorch/pull/71965
[4] https://pkgbuild.vdwaa.nl/?q=-D_GLIBCXX_ASSERTIONS&i=nope&files=&excludeFiles=&repos=

**Comment by CoelacanthusHex** [comment]
AFAIK, the openssl-sys crate is the most common Rust dependency that will be broken when LTO is enabled.

**Reply by yan12125** [reply]
Thanks for sharing! I noticed libgit2-sys is also broken with LTO (first victims!)

https://build.archlinuxcn.org/imlonghao-api/pkg/wezterm-git/log/1655557573
https://build.archlinuxcn.org/imlonghao-api/pkg/broot-git/log/1655557298

/cc @ykelvis @wfxr

**Reply by yan12125** [reply]
Seems rusqlite is also broken with LTO

https://build.archlinuxcn.org/imlonghao-api/pkg/nushell-git/log/1655620791

/cc @axionl

**Reply by yan12125** [reply]
ring is also broken

@ykelvis https://build.archlinuxcn.org/imlonghao-api/pkg/ncspot-git/log/1655652647
@DuckSoft https://build.archlinuxcn.org/imlonghao-api/pkg/shadowsocks-rust-opt-git/log/1655735195 https://build.archlinuxcn.org/imlonghao-api/pkg/shadowsocks-rust-git/log/1655735155

**Reply by yan12125** [reply]
libgit2-sys: @zjuyk https://build.archlinuxcn.org/imlonghao-api/pkg/git-delta-git/log/1655763638
minimp3-sys: @wfxr https://build.archlinuxcn.org/imlonghao-api/pkg/clitrans-git/log/1655789244

I think that's all for frequently-updated packages.

**Comment by lilydjwg** [comment]
The buildarm server had lto enabled and `-Werror=format-security` in the past because I copied a new makepkg.conf and made changes. It now doesn't have these settings, using the `devtools-arm-git` package by @yuyichao.

**Reply by yuyichao** [reply]
Oh I guess I should update the makepkg conf file…

that may take a while though since I lost internet at home due to a storm……

**Reply by yan12125** [reply]
Got it. Probably better to update makepkg.conf there to use the same flags as others.

**Reply by yan12125** [reply]
Sorry to hear that. Hope you get everything well soon.

**Reply by yuyichao** [reply]
Just pushed an update to sync the flags. Should be the same as the flags used on the x86 devtools after the next rebuild run.

(sending from lab... still no internet at home......)

**Comment by yan12125** [comment]
> Rpcs3 doesn't support building with LTO

(From https://build.archlinuxcn.org/imlonghao-api/pkg/rpcs3-git/log/1655587008)

Nice :D

/cc @heavysink

**Reply by heavysink** [reply]
Done.

**Comment by yan12125** [comment]
pcsx2-git fails with

```
/usr/bin/ld: /tmp/cc8PIB7t.ltrans62.ltrans.o: in function `recExitExecution()':
<artificial>:(.text+0x324): undefined reference to `fastjmp_jmp'
/usr/bin/ld: /tmp/cc8PIB7t.ltrans62.ltrans.o: in function `recExecute() [clone .lto_priv.0]':
<artificial>:(.text+0x13355): undefined reference to `fastjmp_set'
```

(From https://build.archlinuxcn.org/imlonghao-api/pkg/pcsx2-git/log/1655616862)

Apparently upstream build system detection is incomplete https://github.com/PCSX2/pcsx2/pull/4938

**Reply by heavysink** [reply]
Done.

**Comment by yan12125** [comment]
First case in archlinuxcn with broken LTO + CGO:

@dctxmei https://build.archlinuxcn.org/imlonghao-api/pkg/clash-dev-git/log/1655647665 https://build.archlinuxcn.org/imlonghao-api/pkg/clash-dctxmei/log/1655647630

**Reply by dctxmei** [reply]
Done

**Comment by yuyichao** [comment]
@lilydjwg It seems that `-fcf-protection` is not supported on arm: https://buildarm.archlinuxcn.org/imlonghao-api/pkg/nheko-git/log/1655784776 which means that devtools-arm-git needs to be updated again https://github.com/yuyichao/devtools-arm/commit/6176c7a37257275f91b2a7001dad8f4c3940fe2e

**Reply by lilydjwg** [reply]
Updated to 0.0.0.11.6176c7a-1.
